### DISCUSSION TITLE
How to calculate Spectral Efficient or Throughput in the Sionna?

### DISCUSSION THREAD

**Discussion by Mio-TW** [NVlabs/sionna#343] [Q&A]
Dear Sionna Team,

We can directly calculate the BLER/BER with the Sionna API.
But I'm curious how to calculate Spectral Efficient or Throughput in the Sionna?
Thank you so much.

**Comment by SebastianCa** [comment]
Hi,
currently, Sionna does not directly support the desired functionality. However, the good news is that implementing these functions should be relatively straightforward. For detailed guidance on this, I recommend checking out [the following paper](https://arxiv.org/pdf/1911.13055.pdf).

If your question refers to the simulation throughput (which I guess is not the case), you might find the [FEC tutorial notebook](https://nvlabs.github.io/sionna/examples/5G_Channel_Coding_Polar_vs_LDPC_Codes.html#Throughput-and-Decoding-Complexity) particularly useful.

I hope that answers your question.

**Reply by 571991380** [reply]
Hi,SebastianCa.I want to simulate the throughput of the entire wireless communication system, rather than the speed of program processing. Can I achieve this using Sionna? Thank you.

**Reply by rwiesmayr** [reply]
Hi @571991380,
Sionna is designed for physical layer simulations. For system-level throughput simulations, dedicated system-level simulators might be more suitable.

Alternatively, you can estimate throughput using Sionna's PUSCH module by:
1. Assuming a bandwidth allocation.
2. Making scheduling assumptions.
3. Running the PUSCH simulation.
4. Inferring system-level throughput from the physical layer results. E.g., one could infer the throughput by `k*(1-BLER)/slot_time`.

For comprehensive system-level analysis, a specialized simulator is recommended.

Hope this helps!

**Comment by 571991380** [comment]
Hi,Mio.Do you konw how to calculate Spectral Efficient or Throughput in the Sionna now?Thank you .

**Reply by Mio-TW** [reply]
Hi @571991380,
As @SebastianCa suggested, I calculated the Tput by self-defined function.
The function is as follows:
```
effective_ofdm_sym = 14 - num_pilot_sym
effective_bits = fft_size * effective_ofdm_sym * num_bits_per_sym * min(TX_ant, RX_ant)
coded_bits = effective_bits * coderate
Tput_bps = coded_bits * 1000 * (subcarrier_spacing/15e3) * (1-BLER)
Tput_Mbps = Tput_bps / (10**6) 
```
For example, 
```
subcarrier_spacing = 15e3
num_pilot_sym = 2
fft_size = 128
num_bits_per_sym = 4
TX_ant = 2
RX_ant = 4
coderate = 0.5
BLER = 0.1

effective_ofdm_sym = 14 - num_pilot_sym = 14 - 2 = 12
effective_bits = fft_size * effective_ofdm_sym * num_bits_per_sym * min(TX_ant, RX_ant) = 128 * 12 * 4 * 2 = 12288
coded_bits = effective_bits * coderate = 12288 * 0.5 = 6144
Tput_bps = coded_bits * 1000 * (1-BLER) = 6144 * 1000 * 1 * (1-0.1) = 5529600
Tput_Mbps = Tput_bps / (10**6) = 5.5296 Mbps
```

You can use it as a reference.

**Reply by 571991380** [reply]
@Mio-TW ,
Thank you for your replay.

I obtain the system bit error rate (BER), the amount of data that the OFDM resource grid can carry "coded_bits" in your code, and then the throughput = coded_bits * (1-BER) / (slot time, 1ms). Is this understanding correct?

**Reply by Mio-TW** [reply]
@571991380,
I think you can directly change the BLER to BER in the function.
Since BER is the number of bit errors divided by the total number of transferred bits, you may need to consider channel coding. 

```
effective_ofdm_sym = 14 - num_pilot_sym
effective_bits = fft_size * effective_ofdm_sym * num_bits_per_sym * min(TX_ant, RX_ant)
coded_bits = effective_bits * coderate
Tput_bps = coded_bits * 1000 * (subcarrier_spacing/15e3) * (1-BER)
Tput_Mbps = Tput_bps / (10**6) 
```
